client/alloc_runner.go

```diff
@@ -103,11 +103,18 @@ func (r *AllocRunner) RestoreState() error {
 
 	// Restore the task runners
 	var mErr multierror.Error
-	for name := range r.taskStatus {
+	for name, status := range r.taskStatus {
 		task := &structs.Task{Name: name}
 		restartTracker := newRestartTracker(r.alloc.Job.Type, r.RestartPolicy)
 		tr := NewTaskRunner(r.logger, r.config, r.setTaskStatus, r.ctx, r.alloc.ID, task, restartTracker)
 		r.tasks[name] = tr
+
+		// Skip tasks in terminal states.
+		if status.Status == structs.AllocClientStatusDead ||
+			status.Status == structs.AllocClientStatusFailed {
```

Comment:
Yeah that is what happens. Failed just distinguishes from Dead that the user process never started.